src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11.java

```diff
@@ -158,9 +160,9 @@ public static void loadNative() {
         // and get cryptoki version from there
         if (this.version.major != 2 && this.version.major != 3) {
             try {
-                CK_INFO p11Info = C_GetInfo();
-                this.version.major = p11Info.cryptokiVersion.major;
-                this.version.minor = p11Info.cryptokiVersion.minor;
+                getInfo();
+                this.version.major = pInfo.cryptokiVersion.major;
+                this.version.minor = pInfo.cryptokiVersion.minor;
```

Comment:
Now that you added getInfo(), then the C_GetInfo() could be made private, so that all callers can just leverage getInfo(), right?

Reply:
Right! We will make C_GetInfo private and delete the overload in SynchronizedPKCS11 for the next iteration.

Reply:
We have analyzed this issue further and think that we need one more iteration. It's possible that the constructor of PKCS11 finishes with PKCS11::pInfo in null. In fact, it's very likely to happen and PKCS11::getInfo is not multi-thread safe. The second problem is that it's possible that a SyncrhonizedPKCS11 instance ends up calling a non-synchronized version of PKCS11::C_GetInfo, if we remove the override. The reason why we added PKCS11::getInfo in the first place was because PKCS11::getVersion is not currently available in 17u —where this feature was originally developed— and there were 2 calls to PKCS11::C_GetInfo. We can safely undo the PKCS11::getInfo change and call PKCS11::getVersion in P11SecretKeyFactory::derivePBEKey.